### PR TITLE
Add advanced augmentations: RandomFisheye, RandomElasticTransform, RandomThinPlateSpline, RandomBloxBlur

### PR DESCRIPTION
**Pull request by edgarriba**
### Description

Add the following augmentations:
- [x] `RandomFisheye`
- [x] `RandomElasticTransform`
- [x] `RandomThiPlateSpline`
- [x] `RandomBloxBlur`

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: kornia/augmentation/augmentation.py [hunk 1]
```diff
@@ -1723,3 +1724,80 @@ def apply_transform(
...
+    Args:
+        center_x (torch.Tensor): Ranges to sample respect to x-coordinate center with shape (2,).
+        center_y (torch.Tensor): Ranges to sample respect to y-coordinate center with shape (2,).
+        gamma (torch.Tensor): Ranges to sample for the gamma values respect to optical center with shape (2,).
+        return_transform (bool): if ``True`` return the matrix describing the transformation applied to each
+            input tensor. If ``False`` and the input is a tuple the applied transformation wont be concatenated.
+        same_on_batch (bool): apply the same transformation across the batch. Default: False.
+        p (float): probability of applying the transformation. Default value is 0.5.
+
+    Examples:
```
Comment (shijianjian) [inline kornia/augmentation/augmentation.py:1741]: Not related, but it will be nice if we can print some images whilst build the docs to show what is accepted and expected.

File: kornia/augmentation/augmentation.py [hunk 2]
```diff
@@ -1723,3 +1724,80 @@ def apply_transform(
...
+
+        if len(data.shape) != 1 and data.shape[0] != 2:
+            raise ValueError(f"Tensor must be of shape (2,). Got: {data.shape}.")
+
+        return data
+
+    def generate_parameters(self, shape: torch.Size) -> Dict[str, torch.Tensor]:
+        B, _, _, _ = shape  # batch_size
+        center_x = self.dist(self.center_x[:1], self.center_x[1:]).rsample(shape[:1])
+        center_y = self.dist(self.center_y[:1], self.center_y[1:]).rsample(shape[:1])
```
Comment (shijianjian) [inline kornia/augmentation/augmentation.py:1780]: We should pre-build the distribution modules and only call resample in this function, to avoid creating distribution modules on the fly.

**Review by shijianjian — CHANGES_REQUESTED** [review]

File: kornia/augmentation/augmentation.py (hunk 2 above)
Comment (edgarriba) [reply kornia/augmentation/augmentation.py:1780]: Ok, we can do it with the new API

Comment (shijianjian) [reply kornia/augmentation/augmentation.py:1780]: We might init two uniform distributions from 0-1 in the initializer. Then scale it to the proper range in the param generator.

File: kornia/augmentation/augmentation.py (hunk 1 above)
Comment (edgarriba) [reply kornia/augmentation/augmentation.py:1824]: let's target this for the next feature release. Wanna go for it with a script to automate this ?
